sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
 	},
 };
 
-struct snd_soc_dai_link dailink[] = {
-	{
-		.name = "SDW0-Playback",
-		.id = 0,
-		.init = headset_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw0_pin2, sdw0_codec, platform),
-	},
-	{
-		.name = "SDW0-Capture",
-		.id = 1,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw0_pin3, sdw0_codec, platform),
-	},
-	{
-		.name = "SDW1-Playback",
-		.id = 2,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw1_pin2, sdw1_codec, platform),
-	},
-	{
-		.name = "SDW3-Capture",
-		.id = 4,
-		.no_pcm = 1,
-		.dpcm_capture = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw3_pin2, sdw3_codec, platform),
-	},
+static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+{
+	const struct snd_soc_acpi_link_adr *link;
+	int num = 0;
+
+	link = mach_params->links;
+	while (link->num_adr) {
+		int part_id;
+
+		part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+		/* playback & capture */
+		if (part_id == 0x711)
+			num++;
+
+		num++;
+		link++;
+	}
+
+	return num;
+}
+
+static void init_dai_link(struct snd_soc_dai_link *dai_links, int id,
+			  char *name,
+			  struct snd_soc_dai_link_component *cpus,
+			  char *cpu_name,
+			  struct snd_soc_dai_link_component *codecs,
+			  int codecs_num)
+{
+	dai_links[id].id = id;
+	dai_links[id].name = name;
+	dai_links[id].platforms = platform_component;
+	dai_links[id].num_platforms = ARRAY_SIZE(platform_component);
+	dai_links[id].nonatomic = true;
+	dai_links[id].no_pcm = 1;
+	dai_links[id].cpus = &cpus[id];
+	dai_links[id].num_cpus = 1;
+	dai_links[id].cpus->dai_name = cpu_name;
+	dai_links[id].codecs = codecs;
+	dai_links[id].num_codecs = codecs_num;
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	unsigned int part_id, unique_id, mfg_id, link_id, class_id;
+	struct snd_soc_dai_link_component *codec;
+	int i, j, dai_num = 1;
+
+	codec = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component) *
+			     link->num_adr, GFP_KERNEL);
```

Comment:
need to test if link != NULL (either here or in the caller)

Reply:
thanks, I will refine it

Reply:
thanks , I will refine it

Reply:
how about using `devm_kcalloc(dev, link->num_adr, sizeof(struct snd_soc_dai_link_component), GFP_KERNEL)`?